File: Project_Graph & Tree_Traversing_Problems/Bài_toán_5.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(int u, vector<vector<int>>& adj, vector<bool>& visited) {
    visited[u] = true;
    for (int v : adj[u]) {
        if (!visited[v])
            dfs(v, adj, visited);
    }
}

bool is_connected(vector<vector<int>>& adj, int n) {
    vector<bool> visited(n, false);
    dfs(0, adj, visited);
    for (bool v : visited)
        if (!v) return false;
    return true;
}

bool is_tree(vector<vector<int>>& adj, int n) {
    vector<bool> visited(n, false);
    vector<int> parent(n, -1);

    function<bool(int)> dfs_tree = [&](int u) {
        visited[u] = true;
        for (int v : adj[u]) {
            if (!visited[v]) {
                parent[v] = u;
                if (!dfs_tree(v)) return false;
            } else if (parent[u] != v)
                return false;
        }
        return true;
    };

    if (!dfs_tree(0)) return false;
    for (bool v : visited)
        if (!v) return false;
    return true;
}

int count_components(vector<vector<int>>& adj, int n) {
    vector<bool> visited(n, false);
    int count = 0;
    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            dfs(i, adj, visited);
            ++count;
        }
    }
    return count;
}

int main() {
    vector<vector<int>> adj = {
        {1}, {0, 2}, {1}, {}
    };
    int n = 4;

    cout << "Connected: " << (is_connected(adj, n) ? "Yes" : "No") << "\n";
    cout << "Is Tree: " << (is_tree(adj, n) ? "Yes" : "No") << "\n";
    cout << "Components: " << count_components(adj, n) << "\n";
    return 0;
}

File: Project_Graph & Tree_Traversing_Problems/Bài_toán_5.py
def dfs(adj, visited, u):
    visited[u] = True
    for v in adj[u]:
        if not visited[v]:
            dfs(adj, visited, v)

def is_connected(adj, n):
    visited = [False] * n
    dfs(adj, visited, 0)
    return all(visited)

def is_tree(adj, n):
    visited = [False] * n
    parent = [-1] * n

    def dfs_tree(u):
        visited[u] = True
        for v in adj[u]:
            if not visited[v]:
                parent[v] = u
                if not dfs_tree(v):
                    return False
            elif parent[u] != v:
                return False
        return True

    if not dfs_tree(0):
        return False
    return all(visited)

def count_components(adj, n):
    visited = [False] * n
    count = 0
    for i in range(n):
        if not visited[i]:
            dfs(adj, visited, i)
            count += 1
    return count

adj = {
    0: [1], 1: [0, 2], 2: [1], 3: []
}
n = 4
print("Connected:", is_connected(adj, n))
print("Is Tree:", is_tree(adj, n))
print("Components:", count_components(adj, n))
